README.md

```diff
@@ -55,6 +55,12 @@ this tool:
 Multiple file names may be comma-separated. Use this to communicate with an installation
 of Apigee Edge that uses a custom certificate for API calls.
 
+`--keyfile -K`
+(optional) The name of the PEM file that represents the private key in a mutual auth connection.
+
+`--certfile -C`
+(optional) The name of the PEM file that represents the certificate in a mutual auth connection.
+
```

Comment:
It would be slick to support password-protected key files, but we can always add this later.  Just a thought.

Reply:
@whitlockjc Let me know if this is sufficient proof of operation. In summary, I've validated I'm able to use the Apigee SDK, which we have wrapped in a tool.  The purpose of the tool is to provide 'context' to developers working on Apigee proxies, or issue other non-proxy targeted Apigee management API calls, e.g. environment KVMs. The wrapping is accomplished through a module called sdk-cmd.js, Key portions of the code are:
```
var sdkCmd = function (options) {
  envSpecified = Boolean(options.env);
  var cmd = mapCmd(options._name);
  options['token'] =   options['token'] = getToken(options); // get the access token, if undefined returned, use UN/PW
  opts.setOptions(options);
```
where all SDK commands are wrapped in a single function that maps the SDK command name to the action (program _name) from the commander interface module. `cmd = mapCmd(options._name) `  hits a switch statement that sets flags as to the input values needed.  In the case of status, which I'm demoing for proof, it does the following:
```
    case 'status':
      needProxy = envSpecified ? false : true;
      needRev = false;
      return sdk.listDeployments;
```
Then the command is invoked.  This also is wrapped, so that if the access token retrieved earlier was in need of a refresh, a chained call is made to get a fresh token followed by the desired SDK call.  Further, since the Javascript promises used by the SDK are an earlier variant, the wrapping supports promise syntax for both our SDK and non-SDK API calls.
```
    opts.call('sdk', cmd, opts,
      function (result) {
        // success
        if (options._name === 'createKVM' || options._name === 'deleteKVM' || options._name === 'createKVMEntry' || options._name === 'deleteKVMEntry')
          logPriviledgedOperation(result, options._name, opts.username, needProxy ? opts.api : `Env KVM: ${opts.environment}`);
        process.stdout.write(`${JSON.stringify(result, null, 2)}\n`);
      }, function (err) {
        // failed
        process.stderr.write(`${err}\n`);
        // Apigee returns 400 when deploying an already deployed proxy. To avoid raising an error
        // in a CI pipleline, the error condition is suppressed and an exit code of 0 returned.
        // err must be cast as a String to access the .match method
        if (options._name === 'deploy' && String(err).match(/already deployed/) ) { 
          process.exit(0);
        } else {
          // the Apigee error response contains the status code, in the 4xx to 5xx range.
          // Extract and subtract 300, as exit status codes are limited to 0 - 255.  So,
          // 4xx codes show as 1xx in the exit code and 5xx errors are 2xx
          var exitCode = /\d+/.exec(err);
          process.exit(exitCode ? Number(exitCode[0]) - 300 : 1);
        }
      });
```
This is the basic setup, now on to the demonstration of the enhancement.

Our company protects our internal API proxy endpoints by securing the Apigee endpoint via mutual auth.  At Apigee the endpoint is api-int.nationwide.com (or in the case of our test environment api-int-test.nationwide.com).  These endpoints are known inside our network as api-int.nwie.net and api-int-test.nwie.net and are proxied to the Apigee endpoints with mutual auth.  This prevents any internet traffic that would hit the api-int(-test).nationwide.com endpoints, unless they can provide the appropriate key/cert.  We had a need for one IT team, using cloud DevOps tools, to hit our internal -test endpoint from the internet.  Here is the demo.

### Invoking the SDK with debug enabled; existing functionality
You'll note that the request options object does not contain any key/cert and it's accessing the api-int-test.nwie.net (internal) proxy endpoint.  It routes through a pass through proxy for logging purposes, but the `v1/o/nationwide-test/apis/ApigeeProxy_1/deployments` is the Apigee API for reporting a proxy's deployments.
```
$ nw-apigee-cli status -D                                                                                                                                                                                                                      
listdeployments: {"header":["client_id:boO3e8CYA8zZspv2E1VR20cPubooDr6p","X-NW-Message-ID:shompek2-1568209130974"],"insecure":true,"token":"Apigee identity token (JWT) ","username":"shompek2@nationwide.com","debug":true,"baseuri":"https://api-int-test
.nwie.net/utility/apigee-proxy","import-only":true,"api":"ApigeeProxy_1","organization":"nationwide-test","revision":1,"cliUri":"https://api-int-test.nwie.net/utility/apigee-proxy/v1","request":{"headers":{"Authorization":"Bearer JWT","client_id":"boO3e8CYA8zZspv2E1VR20cPubooDr6p","X-NW-Message-ID":"shompek2-1568209130974"},"json":true,"agentOptions":{"rejectUnauthorized":false}},"asynclimit":4,"prompt":true}                                                     
Going to invoke "https://api-int-test.nwie.net/utility/apigee-proxy/v1/o/nationwide-test/apis/ApigeeProxy_1/deployments"                                                                                                                       
List of deployed APIs: {"environment":[{"name":"test","revision":[{"configuration":{"basePath":"/","configVersion":"SHA-512:dcc036eebb508fb678262f67ccc1376f92a5d35186402bd557fe32e898dc6aeb5bd6f6d74af396acf8d456c252a398beb361cde516783bc7624
529eb6bb725fa","steps":[]},"name":"4","server":[{"pod":{"name":"rea1mp302-7","region":"us-east-1"},"status":"deployed","type":["message-processor"],"uUID":"b9dc7357-682c-4084-a803-e6f785490442"},{"pod":{"name":"rea1mp302-7","region":"us-ea
st-1"},"status":"deployed","type":["message-processor"],"uUID":"6c97fe1f-9996-4aa7-9c5a-3a765b397c5a"},{"pod":{"name":"rea1mp302-7","region":"us-east-1"},"status":"deployed","type":["message-processor"],"uUID":"1524109a-b85a-460f-84c3-600f
822f35f4"},{"pod":{"name":"rea1mp302-7","region":"us-east-1"},"status":"deployed","type":["message-processor"],"uUID":"ec2a9b93-ecf1-46fc-af15-f4473a7a3f56"},{"pod":{"name":"rea1rt302-0","region":"us-east-1"},"status":"deployed","type":["r
outer"],"uUID":"dc678fad-555b-4f5d-bab7-bc4ac478532a"},{"pod":{"name":"rea1rt302-0","region":"us-east-1"},"status":"deployed","type":["router"],"uUID":"e2bf78e9-38d1-400e-b9de-1d54215e4342"},{"pod":{"name":"rea1rt302-0","region":"us-east-1
"},"status":"deployed","type":["router"],"uUID":"95ed91be-1d80-46a9-982d-8528975a8392"},{"pod":{"name":"rea1rt302-0","region":"us-east-1"},"status":"deployed","type":["router"],"uUID":"2c4a7dc4-0df0-44f5-b6bb-b44d25731737"},{"pod":{"name":
"rea1rt302-0","region":"us-east-1"},"status":"deployed","type":["router"],"uUID":"d8153d5e-6c12-4a01-ba78-19ebb499e5ab"},{"pod":{"name":"rea1rt302-0","region":"us-east-1"},"status":"deployed","type":["router"],"uUID":"27f75417-bb6e-4f05-8c
5b-825c0ff39a21"}],"state":"deployed"}]}],"name":"ApigeeProxy_1","organization":"nationwide-test"}                                                                                                                                             
All done                                                                                                                                                                                                                                       
{                                                                                                                                                                                                                                              
  "deployments": [                                                                                                                                                                                                                             
    {                                                                                                                                                                                                                                          
      "name": "ApigeeProxy_1",                                                                                                                                                                                                                 
      "environment": "test",                                                                                                                                                                                                                   
      "revision": 4,                                                                                                                                                                                                                           
      "state": "deployed",                                                                                                                                                                                                                     
      "basePath": "/"                                                                                                                                                                                                                          
    }                                                                                                                                                                                                                                          
  ]                                                                                                                                                                                                                                            
}                                                                                                                                                                                                                                              
```
### Invoking the SDK with debug enabled; external endpoint
Here I set the environment variable `CLI_PROXYDOMAIN=nationwide.com`, which causes the request to hit the external proxy endpoint.
```
$ export CLI_PROXYDOMAIN=nationwide.com
$ nw-apigee-cli status -D
listdeployments: {"header":["client_id:boO3e8CYA8zZspv2E1VR20cPubooDr6p","X-NW-Message-ID:shompek2-1568209779682"],"insecure":true,"token":"Apigee identiy token (JWT)","username":"shompek2@nationwide.com","debug":true,"baseuri":"https://api-int-test.nationwide.com/utility/apigee-proxy","import-only":true,"api":"ApigeeProxy_1","organization":"nationwide-test","revision":1,"cliUri":"https://api-int-test.nationwide.com/utility/apigee-proxy/v1","request":{"headers":{"Authorization":"Bearer JWT","client_id":"boO3e8CYA8zZspv2E1VR20cPubooDr6p","X-NW-Message-ID":"shompek2-1568209779682"},"json":true,"agentOptions":{"rejectUnauthorized":false}},"asynclimit":4,"prompt":true}
Going to invoke "https://api-int-test.nationwide.com/utility/apigee-proxy/v1/o/nationwide-test/apis/ApigeeProxy_1/deployments"
Error: HTTP error 400

$ echo $?
100
```
Here we see the request was blocked and the SDK reports a 400 error

### Invoking the SDK with debug enabled; external endpoint
Now I add the key and cert files to the request (sanitized for privacy), and it succeeds, even though it's hitting our external endpoint which requires mutual auth to connect.
```
$ export CLI_KEY_FILE=certs/key.pem
$ export CLI_CERT_FILE=certs/cert.pem

$ nw-apigee-cli status -D
listdeployments: {"header":["client_id:boO3e8CYA8zZspv2E1VR20cPubooDr6p","X-NW-Message-ID:shompek2-1568210172074"],"insecure":true,"token":"Apigee identify token (JWT)","username":"shompek2@nationwide.com","debug":true,"baseuri":"https://api-int-test.nationwide.com/utility/apigee-proxy","import-only":true,"api":"ApigeeProxy_1","organization":"nationwide-test","revision":1,"cliUri":"https://api-int-test.nationwide.com/utility/apigee-proxy/v1","request":{"headers":{"Authorization":"Bearer JWT","client_id":"boO3e8CYA8zZspv2E1VR20cPubooDr6p","X-NW-Message-ID":"shompek2-1568210172074"},"json":true,"agentOptions":{"rejectUnauthorized":false},"key":{"type":"Buffer","data":[... key file data ... secret ...]},"cert":{"type":"Buffer","data":[... cert file data ... secret ...]}},"keyfile":"../explore-pet/certs/key.pem","certfile":"../explore-pet/certs/cert.pem","asynclimit":4,"prompt":true}
Going to invoke "https://api-int-test.nationwide.com/utility/apigee-proxy/v1/o/nationwide-test/apis/ApigeeProxy_1/deployments"
List of deployed APIs: {"environment":[{"name":"test","revision":[{"configuration":{"basePath":"/","configVersion":"SHA-512:dcc036eebb508fb678262f67ccc1376f92a5d35186402bd557fe32e898dc6aeb5bd6f6d74af396acf8d456c252a398beb361cde516783bc7624529eb6bb725fa","steps":[]},"name":"4","server":[{"pod":{"name":"rea1mp302-7","region":"us-east-1"},"status":"deployed","type":["message-processor"],"uUID":"b9dc7357-682c-4084-a803-e6f785490442"},{"pod":{"name":"rea1mp302-7","region":"us-east-1"},"status":"deployed","type":["message-processor"],"uUID":"6c97fe1f-9996-4aa7-9c5a-3a765b397c5a"},{"pod":{"name":"rea1mp302-7","region":"us-east-1"},"status":"deployed","type":["message-processor"],"uUID":"1524109a-b85a-460f-84c3-600f822f35f4"},{"pod":{"name":"rea1mp302-7","region":"us-east-1"},"status":"deployed","type":["message-processor"],"uUID":"ec2a9b93-ecf1-46fc-af15-f4473a7a3f56"},{"pod":{"name":"rea1rt302-0","region":"us-east-1"},"status":"deployed","type":["router"],"uUID":"dc678fad-555b-4f5d-bab7-bc4ac478532a"},{"pod":{"name":"rea1rt302-0","region":"us-east-1"},"status":"deployed","type":["router"],"uUID":"e2bf78e9-38d1-400e-b9de-1d54215e4342"},{"pod":{"name":"rea1rt302-0","region":"us-east-1"},"status":"deployed","type":["router"],"uUID":"95ed91be-1d80-46a9-982d-8528975a8392"},{"pod":{"name":"rea1rt302-0","region":"us-east-1"},"status":"deployed","type":["router"],"uUID":"2c4a7dc4-0df0-44f5-b6bb-b44d25731737"},{"pod":{"name":"rea1rt302-0","region":"us-east-1"},"status":"deployed","type":["router"],"uUID":"d8153d5e-6c12-4a01-ba78-19ebb499e5ab"},{"pod":{"name":"rea1rt302-0","region":"us-east-1"},"status":"deployed","type":["router"],"uUID":"27f75417-bb6e-4f05-8c5b-825c0ff39a21"}],"state":"deployed"}]}],"name":"ApigeeProxy_1","organization":"nationwide-test"}
All done
{
  "deployments": [
    {
      "name": "ApigeeProxy_1",
      "environment": "test",
      "revision": 4,
      "state": "deployed",
      "basePath": "/"
    }
  ]
}
```

Let me know if you require something more as evidence of correct operation.  This example is listDeployments, but I've successfully validated deployExistingRevision, undeploy, and the various KVM commands.

Reply:
I just meant it would be nice to have tests that the options were set as necessary, like you see here: https://github.com/apigee/apigeetool-node/blob/master/test/testoptions.js  We don't have end-to-end tests and I don't think we need to make this the first one.  Just add tests for the CLI options being set/unset as expected, and we'll get this merged.  I'm sorry this wasn't clear before.